### ISSUE TITLE
Fix S2955 FP/FN: False positive with nullable parameters

### ISSUE THREAD

**Issue by AHarman** [SonarSource/sonar-dotnet#8933]
### Description

S2955 gives a false positive when using nullable arguments for a method.

### Repro steps

```c#
public static bool BothNullOrNeitherNull<T1, T2>(T1? input1, T2? input2)
{
    if (input1 != null && input2 == null)
    {
        return false;
    }
    else if (input1 == null && input2 != null)
    {
        return false;
    }
    return true;
}
```

A warning appears for each null comparison

### Expected behavior

No message - `T1?` always has a default value of `null`, even if  `T1` is a value type. 

### Actual behavior

A warning appears with the text "Use a comparison to 'default(T1)' instead or add a constraint to 'T1' so that it can't be a value type."

### Known workarounds

None that I know of

### Related information

* C#/VB.NET Plugins version: SonarLint for Visual Studio 7.3.0.77872
* Visual Studio version: 17.9.2
* dotnet version: 8
* Operating System: Windows 10

**Comment by jilles-sg** [comment]
It is somewhat confusing, but the rule is right. If `T` is a type parameter that may be both a value or a reference type, `T?` is the same as `T` at run time. Reference: https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#generics

For example, given `void f<T>(T? x) { }`, `f<int>(null);` does not build, but adding `where T: struct` to the method definition makes it build.

By the way, I do not agree with the concept of the rule since it may prevent the programmer from doing what needs to be done. Given `T x` with `T` unconstrained, `x == null` means to check for null with the understanding that this can never be true if `T` is a non-nullable value type. Making this more explicit using `typeof(T).IsValueType` may not be satisfactory since nullable value types are still value types. Also, it is hard to call a more constrained from a less constrained method, even if the constraint is known to be true at run time.

**Comment by gregory-paidis-sonarsource** [comment]
Hey there!
Thanks for the feedback, it helps a lot.

I discussed it a bit internally with my team, and I am not sure if we should change anything here.
As @jilles-sg very astutely pointed out, the rule is working as intended, but nullable generics are behaving in a bit of a counter-intuitive manner.

Based on the fact that `T?` is `T` is `T` is a value type, this is a true positive.

About the second suggestion, I think that the rule is pointing in the correct direction, making the user think of value vs reference types. 

Please let me know if you have more feedback so that I can close this issue.

**Comment by jilles-sg** [comment]
Looking at our codebase, I see various cases where such a comparison serves only to prevent null pointer exceptions, not to add functionality. Perhaps the rule can be silenced by using, for example, `x?.ToString()` instead of guarding `x.ToString()` with `x != null`? In other cases, the rule is right and the code implicitly expects the `value == null` case to be possible.

Apart from that, the recommendation for `object.Equals(value, default(T))` is a needless pessimization. It is better to use `EqualityComparer<T>.Default.Equals(value, default)` instead, since that uses an `IEquatable<T>` implementation if it is available and has special cases for enums and nullable value types. This avoids boxing in case `T` is a value type.

**Comment by gregory-paidis-sonarsource** [comment]
Hey again @jilles-sg .
About your suggestion to avoid boxing, it's a really good point.
I already [made a PR](https://github.com/SonarSource/rspec/pull/3794/files) to change the suggestion on the rule, to be more performant in case `T` is a value type.
Thanks a lot!
